node/lib/util/push.js

```diff
@@ -46,6 +46,113 @@ const SubmoduleConfigUtil = require("./submodule_config_util");
+ * @param {NodeGit.Commit}     commit
+ */
+exports.getPushMap = co.wrap(function*(repo, remoteName, source, target,
+                                       commit) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.isString(remoteName);
+    assert.isString(source);
+    assert.isString(target);
+    assert.instanceOf(commit, NodeGit.Commit);
+
+    const submoduleSet = new Set(yield SubmoduleUtil.listOpenSubmodules(repo));
+    for (const s of yield SubmoduleUtil.listAbsorbedSubmodules(repo)) {
+        submoduleSet.add(s);
+    }
+    const submodules = [...submoduleSet];
+    const pushMap = yield SubmoduleUtil.getSubmoduleShasForCommit(repo,
+                                                                  submodules,
+                                                                  commit);
+    const bareRepos = {};
+    for (const sub of Object.keys(pushMap)) {
+        const subBareRepo = yield SubmoduleUtil.getBareRepo(repo, sub);
+        bareRepos[sub] = subBareRepo;
+    }
+
+    const trackingBranches = new Set();
+    trackingBranches.add(`refs/remotes/${remoteName}/${target}`);
+    let tracking;
+    try {
+        const sourceBranch = yield repo.getReference(source);
+        tracking = yield GitUtil.getTrackingInfo(repo, sourceBranch);
+    } catch (e) {
+        // we have no local branch? maybe source is a sha.
+        tracking = null;
+    }
+
+    if (tracking !== null) {
+        if (tracking.remoteName !== null) {
+            trackingBranches.add(
+                `refs/remotes/${tracking.remoteName}/${target}`);
+        }
+        if (tracking.pushRemoteName !== null) {
+            trackingBranches.add(
+                `refs/remotes/${tracking.pushRemoteName}/${target}`);
+        }
+    }
+
+    for (const branch of trackingBranches) {
+        let reference = null;
+        try {
+            reference = yield NodeGit.Reference.lookup(repo, branch);
+        } catch (e) {
+            // if this ref doesn't exist, OK
+            continue;
+        }
+        const id = reference.target();
+        const trackingCommit = yield NodeGit.Commit.lookup(
+            repo, id);
+        const getter = SubmoduleUtil.getSubmoduleShasForCommit;
+        const submoduleShasForCommit = yield getter(repo, Object.keys(pushMap),
+                                                    trackingCommit);
+        for (const sub of Object.keys(pushMap)) {
+            const sha = submoduleShasForCommit[sub];
+            const subRepo = bareRepos[sub];
+            const commitToPush = yield subRepo.getCommit(pushMap[sub]);
+            let trackingCommit;
+            try {
+                trackingCommit = yield subRepo.getCommit(sha);
+            } catch (e) {
+                // We haven't fetched this commit in this submodule,
+                // so we can't do an ancestry check
+                continue;
+            }
+            if (NodeGit.Graph.descendantOf(subRepo, commitToPush,
```

Comment:
Do you need to check if it's descendantOf or equal to? If I'm wrong, ignore me, but if I'm right, there should probably be a test case that fails because this is missing.